indra/db/populate_content.py

```diff
@@ -0,0 +1,705 @@
+        for pmid in pmids_to_add:
+            pmid_data = article_info[pmid]
+            rec = (
+                pmid, pmid_data.get('pmcid'), 
+                self.fix_doi(pmid_data.get('doi')),
+                pmid_data.get('pii')
+                )
+            text_ref_records.append(
+                tuple([None if not r else r for r in rec])
+                )
+            abstract = pmid_data.get('abstract')
+            # Make sure it's not an empty or whitespace-only string
+            if abstract and abstract.strip():
+                abstract_gz = zip_string(abstract)
+                text_content_info[pmid] = \
+                    (self.my_source, 'text', texttypes.ABSTRACT, abstract_gz)
+
+        db.copy(
+            'text_ref', 
+            text_ref_records, 
+            ('pmid', 'pmcid', 'doi', 'pii',)
+            )
+
+        # Build a dict mapping PMIDs to text_ref IDs
+        pmid_list = list(text_content_info.keys())
+        tref_list = db.select(
+            'text_ref', 
+            db.TextRef.pmid.in_([p for p in pmid_list])
+            )
+        pmid_tr_dict = {pmid:trid for (pmid, trid) in 
+                        db.get_values(tref_list, ['pmid', 'id'])}
+
+        # Add the text_ref IDs to the content to be inserted
+        text_content_records = []
+        for pmid, tc_data in text_content_info.items():
+            if pmid not in pmid_tr_dict.keys():
+                continue
+            tr_id = pmid_tr_dict[pmid] 
+            text_content_records.append((tr_id,) + tc_data)
+
+
+        db.copy(
+            'text_content', 
+            text_content_records,
+            cols=('text_ref_id', 'source', 'format', 'text_type', 'content',)
+            )
+        return True
+
+
+    def populate(self, db, n_procs = 1, continuing=False):
+        """
+        Perform the initial population of the pubmed content into the database.
+        
+        Inputs
+        ------
+        db: indra.db.DatabaseManager instance
+            The database to which the data will be uploaded.
+        n_procs: int
+            The number of processes to use when parsing xmls.
+        continuing: bool
+            If true, assume that we are picking up after an error, or otherwise
+            continuing from an earlier process. This means we will skip over
+            source files contained in the database. If false, all files will be
+            read and parsed.
+        """
+        xml_files = self.get_file_list()
+        tr_list = db.select(
+            'source_file', 
+            db.SourceFile.source==self.my_source
+            )
+        existing_files = [tr.name for tr in tr_list]
+        
+        q = mp.Queue()
+        proc_list = []
+        for xml_file in xml_files:
+            if continuing and xml_file in existing_files:
+                print("Skipping %s. Already uploaded." % xml_file)
+                continue
+            if xml_file not in existing_files:
```

Comment:
Refactor to handle issue of incomplete upload due to error in file already marked as uploaded.